src/position.cpp

```diff
@@ -204,87 +202,206 @@ Position& Position::set(const string& fenStr, bool isChess960, StateInfo* si, Th
-
       else
-          continue;
+          UCI::critical_error(std::string("Invalid FEN. Expected castling rights. Got: ") + std::string(1, token));
+
+      if (piece_on(rsq) != rook)
+          UCI::critical_error("Invalid FEN. Trying to set castling rights without required rook.");
 
       set_castling_right(c, rsq);
   }
 
-  // 4. En passant square.
-  // Ignore if square is invalid or not on side to move relative rank 6.
-  bool enpassant = false;
+  ss >> std::ws;
 
-  if (   ((ss >> col) && (col >= 'a' && col <= 'h'))
-      && ((ss >> row) && (row == (sideToMove == WHITE ? '6' : '3'))))
+  // 4. En passant square. Faux ep-square is ignored. Otherwise invalid ep-square is an error.
+  bool enpassant = false;
+  unsigned char col, row;
+  if (!(ss >> col))
+      UCI::critical_error("Invalid FEN. Unexpected end of stream.");
+  if (col != '-')
   {
-      st->epSquare = make_square(File(col - 'a'), Rank(row - '1'));
-
-      // En passant square will be considered only if
-      // a) side to move have a pawn threatening epSquare
-      // b) there is an enemy pawn in front of epSquare
-      // c) there is no piece on epSquare or behind epSquare
-      enpassant = pawn_attacks_bb(~sideToMove, st->epSquare) & pieces(sideToMove, PAWN)
-               && (pieces(~sideToMove, PAWN) & (st->epSquare + pawn_push(~sideToMove)))
-               && !(pieces() & (st->epSquare | (st->epSquare + pawn_push(sideToMove))));
+      if (!(ss >> row))
+          UCI::critical_error("Invalid FEN. Unexpected end of stream.");
+
+      if (   (col >= 'a' && col <= 'h')
+          && (row == (sideToMove == WHITE ? '6' : '3')))
+      {
+         st->epSquare = make_square(File(col - 'a'), Rank(row - '1'));
+
+         // En passant square will be considered only if
+         // a) side to move have a pawn threatening epSquare
+         // b) there is an enemy pawn in front of epSquare
+         // c) there is no piece on epSquare or behind epSquare
+         enpassant = pawn_attacks_bb(~sideToMove, st->epSquare) & pieces(sideToMove, PAWN)
+                  && (pieces(~sideToMove, PAWN) & (st->epSquare + pawn_push(~sideToMove)))
+                  && !(pieces() & (st->epSquare | (st->epSquare + pawn_push(sideToMove))));
+      }
+      else
+          UCI::critical_error("Invalid FEN. Invalid en-passant square.");
   }
 
   if (!enpassant)
       st->epSquare = SQ_NONE;
 
-  // 5-6. Halfmove clock and fullmove number
-  ss >> std::skipws >> st->rule50 >> gamePly;
+  ss >> std::skipws;
+
+  // 5-6. Halfmove clock and fullmove number. Either none or both must be present.
+  //      If they are not present then values are the same as for startpos.
+  if (ss >> st->rule50)
+  {
+      if (!(ss >> gamePly))
+          UCI::critical_error("Invalid FEN. Unexpected end of stream or invalid numbers.");
+  }
+  else
+  {
+      st->rule50 = 0;
+      gamePly = 1;
+  }
+
+  // Technically, positions with rule50==100 are correct, just no moves can be made further.
```

Comment:
This is quite unnecessary IMO, rule50 count doesn't really affect gameplay, and user can easily artifically bypass it anyways.

Reply:
While the 50-move rule and 75-move rule do apply in over-the-board play, they do not apply for certain purposes in endgame theory.  In particular, endgame tablebases will allow one to step through a forced with that takes far longer, and this would (presulably) result in a larger count here.  Instead of refusing to import the FEN, I think it would be better for Stockfish to clamp rule50 to 150.